sdk/search/Azure.Search/src/Models/IndexDocumentsAction{T}.cs

```diff
@@ -0,0 +1,216 @@
+        /// </summary>
+        public IndexActionType ActionType { get; private set; }
+
+        /// <summary>
+        /// The document to index.
+        /// </summary>
+        public T Document { get; private set; }
+
+        /// <summary>
+        /// Private constructor to prevent direct instantiation.
+        /// </summary>
+        private IndexDocumentsAction() { }
+
+        /// <summary>
+        /// Initializes a new instance of the IndexDocumentsAction class.
+        /// </summary>
+        /// <param name="type">
+        /// The operation to perform on the document.
+        /// </param>
+        /// <param name="doc">The document to index.</param>
+        public IndexDocumentsAction(IndexActionType type, T doc)
+        {
+            Debug.Assert(Enum.IsDefined(typeof(IndexActionType), type));
+
+            ActionType = type;
+            Document = doc;
+        }
+
+        /// <summary>
+        /// Serialize the document write action.
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        void IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
+        {
+            Debug.Assert(writer != null);
+
+            writer.WriteStartObject();
+
+            writer.WritePropertyName("@search.action");
+            writer.WriteStringValue(ActionType.ToSerialString());
+
+            // TODO: XXXXX - Investigate a more efficient way of injecting properties into user models with System.Text.Json
+
+            // HACK: Serialize the user's model, parse it, and then write each
+            // of its properties as if they were our own.
+            string json = JsonSerializer.Serialize<T>(Document, JsonExtensions.SerializerOptions);
+            using JsonDocument nested = JsonDocument.Parse(json);
+            if (nested.RootElement.ValueKind == JsonValueKind.Object)
+            {
+                foreach (JsonProperty property in nested.RootElement.EnumerateObject())
+                {
+                    property.WriteTo(writer);
+                }
+            }
+
+            writer.WriteEndObject();
+        }
+    }
+
+    /// <summary>
+    /// Creates <see cref="IndexDocumentsAction{T}"/> instances to add to an
+    /// <see cref="IndexDocumentsBatch{T}"/> and submit via
+    /// <see cref="SearchIndexClient.IndexDocuments"/>.
+    /// </summary>
+    public static class IndexDocumentsAction
+    {
+        /// <summary>
+        /// Create an <see cref="IndexDocumentsAction{T}"/> to upload.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The .NET type that maps to the index schema. Instances of this type
+        /// can be retrieved as documents from the index.
+        /// </typeparam>
+        /// <param name="document">The document to upload.</param>
+        /// <returns>
+        /// An <see cref="IndexDocumentsAction{T}"/> to upload.
+        /// </returns>
+        public static IndexDocumentsAction<T> Upload<T>(T document) =>
```

Comment:
And be sure to document argument exceptions here as well as the constructor if people will be interacting with these more.

Reply:
Yeah - I'm going to file a tracking bug on getting all the exceptions documented.  There are others, like JsonException, that we need to document globally as well.